### PR TITLE
The algorithm is able to extract the ROIs of the custom sheets

### LINKED ISSUE #104: The Algorithm classifies all the ROI areas correctly.
As of now, Our Application is capable of classifying the difference between Custom Sheets and Scantron Sheets. If the Custom Sheet is founded the application is giving output has "Not Yet Supported". In this issue we are going to change that. Our Scantron.py algorithm works in a way where it first identifies the region of interest ROI for scantron sheet. We want the algorithm to do same for our custom sheet, which would be a milestone for our goal.

Tasks:

- Modifying AI Algorithm to extract the ROI area on custom sheets.
- ROI Areas are:
    - 1st column answers
    - 2nd column answers 
    - Key Id
    - Student Id
   
Developer Note:
You may have to edit BubbleScan-AI Repo.

Success Criteria:

- [x] Effective Extraction of ROI area.
- [x] Differentiating different Key Id.

### PR DESCRIPTION
**Pull request by ramezmosad**
Fixes #104 

**What was changed?**
* Introduced two new classes, SheetProcessor and CustomProcessor, to handle the extraction of Regions of Interest (ROIs) from custom answer sheets.
* Created the SheetProcessor abstract class to provide a base structure for sheet processing, which can be extended to support various sheet types.
* Implemented CustomProcessor, a subclass of SheetProcessor, specifically for extracting ROIs of custom answer sheets.
* Modified Scantron95945 so it inherits SheetProcessor.
* Set up a new folder structure within data/:
     * data/customAligned for storing aligned images extracted from the custom PDF sheets.
     * data/customROIs for saving the ROIs of each custom sheet image in separate subfolders, ensuring that multi-page documents are handled correctly.

**Why was it changed?**
This change addresses the need to properly extract ROIs for each page of custom answer sheets in a structure similar to that used for Scantron sheets.

**How was it changed?**
1. New SheetProcessor Abstract Class:

* Created SheetProcessor as an abstract base class, defining a standard interface for sheet processing. This class can be extended by other sheet processing types, including CustomProcessor and Scantron95945.
* Added reusable methods, like extractImagesFromPdf, extractROIs that allow subclasses to handle PDF image extraction without code duplication.

2. New CustomProcessor Class:

* Implemented CustomProcessor to handle the specific layout and structure of the custom answer sheets.
* Defined methods extractROIs and crop_roi in CustomProcessor to:
    * Extract images from each page of the custom sheet PDF.
    * Crop ROIs for each page in a structured way, saving them in a dedicated folder under customROIs.
* Organized the ROIs by creating subfolders for each image within customROIs.

3. New Folder Structure for Custom Sheets:
* Added customAligned and customROIs folders to data/ to separate custom sheet data from Scantron data. This keeps custom sheet processing independent, organized, and easy to manage.

**Screenshots that show the changes (if applicable):**
Screenshot 2024-10-27 at 3 08 54 AM
Screenshot 2024-10-27 at 3 09 03 AM
Screenshot 2024-10-27 at 3 09 13 AM
Screenshot 2024-10-27 at 3 09 21 AM
Screenshot 2024-10-27 at 3 09 44 AM

### CODE REVIEW

File: BubbleScan-AI/SheetProcessor.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from abc import ABC, abstractmethod
+import os
+import cv2
+import numpy as np
+
```
Comment (yildizahmetali) [inline BubbleScan-AI/SheetProcessor.py:5]: The new SheetProcessor abstract class provides a base for various sheet types, reducing code duplication and making CustomProcessor and Scantron95945 easy to extend.

**Review by yildizahmetali — COMMENTED** [review]
Using Separate folders (customAligned and customROIs) help manage custom sheet data effectively, keeping it distinct from Scantron data. The run_custom_extraction.py script allows straightforward testing, while additional docstrings and exception handling could further enhance clarity and reliability.

Overall, clean, extendable, and well-structured update – great job!

File: BubbleScan-AI/Custom.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            image_path = os.path.join(aligned_images_folder, image_file)
+            print(f"Processing file: {image_file} for ROI extraction")
+
+            # Create a unique folder for each image's ROIs in customROIs
+            image_name = os.path.splitext(image_file)[0]
+            roi_folder = os.path.join(roi_folder_base, image_name)
+            os.makedirs(roi_folder, exist_ok=True)
+
+            # Extract and save ROIs based on updated layout
+            first_column, second_column, key_id, student_id = self.crop_roi(image_path, roi_folder)
```
Comment (mylescorrea) [inline BubbleScan-AI/Custom.py:90]: nice job splitting the first column and key id since they are not entirely aligned

**Review by mylescorrea — COMMENTED** [review]
Tried this on a separate customSheets randomly chosen by the 50 the class marked up and it correctly separated the columns! Great job can't wait to see when the ai scan is implemented in these!

File: BubbleScan-AI/tests/run_custom_extraction.py
Comment (Chirag2x) [inline BubbleScan-AI/tests/run_custom_extraction.py:1]: I liked it the way you use the test case for everyone to verify the working of the ROI's of sheets.

**Review by Chirag2x — APPROVED** [review]
You have done a good Job @ramezmosad, This satisfy all the success criteria. I noticed you are using different file for the custom sheet identification rather using the same Scantron.py file. So are you going to switch the files based on the input of the user. Like if user select's the custom sheets then you are going to use that file by calling the function or you have some other plans? If you are gonna using the function calling then you also have to do error handling for mismatch files. It would be really great if you start updating the documentation for the same which is presented in BubbleScan-AI folder.

**Comment by ramezmosad** [comment]
@Chirag2x Yes, I feel like it would be easier to have two separate files for handling the two different sheet types. And yeah different functions would be called based one whichever sheet type the user chooses. Do you think having it all in one file would be easier? I was just unsure about how I would go about that. There's the option of just implementing the methods that are in Custom.py inside Scantron.py. For example, inside Scantron.py I would add an "extractCustomROIs" method and that would just be the "extractROIs" method that's defined in Custom.py. I'm not sure which way makes more sense so if you have any recommendations please let me know!

**Comment by Chirag2x** [comment]
Well, Both of them are right approach. Let go ahead with the method you pushed. I am approving the merge.

**Pull request merged**